go/felix/intdataplane/floating_ip_mgr.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intdataplane
+
+import (
+	log "github.com/Sirupsen/logrus"
+	"github.com/projectcalico/felix/go/felix/iptables"
+	"github.com/projectcalico/felix/go/felix/proto"
+	"github.com/projectcalico/felix/go/felix/rules"
+	"reflect"
+)
+
+// A floating IP is an IP that can be used to reach a particular workload endpoint, but that the
+// endpoint itself is not aware of.  The 'floating IP' terminology comes from OpenStack, but the
+// concept can be useful with workload orchestration platforms more generally.  OpenStack
+// installations use floating IPs for two reasons: (1) IP mobility, aka as a 'service IP' - i.e. to
+// have an IP that can be moved from time to time to target a different workload, for example,
+// depending on which of those workloads is currently the 'active' one for some HA service; (2) to
+// allow a particular workload to be reachable (inbound) from the Internet, when the networking
+// driver normally doesn't allow an Internet-routable IP to be assigned as a workload's primary IP.
+// In Calico (2) is irrelevant, because a workload can have an Internet-routable IP as its primary
+// IP, but the idea of service IP mobility is still useful, and floating IPs are one way of
+// providing mobile service IPs.
+//
+// Implementation-wise, a floating IP is simply a destination IP address - the 'external' IP - that
+// gets DNAT'd by the compute node to the workload's normal IP address - the 'internal' IP - on
+// packets that are sent to the workload.
+//
+// There generally _isn't_ a corresponding SNAT for the other direction; in other words, packets
+// sent _from_ a workload don't have their source address changing to a floating IP, if the workload
+// has floating IPs associated with it.  This is because people don't expect floating IPs to do
+// that, and its faster in the datapath to avoid NATs if we can, especially when sending to another
+// workload nearby.  (And for sending outbound to the Internet, there will be an SNAT if needed at
+// the data center's border gateway.)  However SNAT is needed in the loopback case where a workload
+// sends to a floating IP that maps back to itself; for this case the datapath processing must be as
+// follows:
+//
+// 1. Workload sends to <floating IP>, so packet has SRC=<workload IP> DST=<floating IP>
```

Comment:
I think the lack of a `.` at the end of these lines will render them as headings in godoc, which is probably not what you want.